oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/token/JwtOAuth2TokenBuilder.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.server.authorization.token;
+
+import java.util.Map;
+import java.util.function.Consumer;
+
+import org.springframework.security.oauth2.jose.JoseHeader;
+import org.springframework.security.oauth2.jose.jws.SignatureAlgorithm;
+import org.springframework.security.oauth2.jwt.Jwt;
+import org.springframework.security.oauth2.jwt.JwtClaimsSet;
+import org.springframework.security.oauth2.jwt.JwtEncoder;
+
+/**
+ * A token builder for JWT tokens. Allows to customize claims and headers of a JWT token.
+ *
+ * @see JwtClaimsSet
+ * @see JoseHeader
+ *
+ * @author Alexey Nesterov
+ * @since 0.1.0
+ */
+public class JwtOAuth2TokenBuilder implements OAuth2TokenBuilder<Jwt> {
+
+	private final JoseHeader.Builder joseHeaderBuilder;
+	private final JwtClaimsSet.Builder jwtClaimsSetBuilder;
+	private final JwtEncoder jwtEncoder;
+
+	private JwtOAuth2TokenBuilder(JwtEncoder jwtEncoder) {
+		this.jwtEncoder = jwtEncoder;
+		this.joseHeaderBuilder = JoseHeader.withAlgorithm(SignatureAlgorithm.RS256);
+		this.jwtClaimsSetBuilder = JwtClaimsSet.builder();
+	}
+
+	/**
+	 * Create a new token builder for JWT token with specified encoder.
+	 * @param jwtEncoder the JWT encoder
+	 * @return new {@code JwtOAuth2TokenBuilder}
+	 */
+	public static JwtOAuth2TokenBuilder withEncoder(JwtEncoder jwtEncoder) {
+		return new JwtOAuth2TokenBuilder(jwtEncoder);
+	}
+
+	/**
+	 * Claims set customizer. The given consumer will be invoked with pre-configured set of claims.
+	 * @param  claimsCustomizer a {@code JwtClaimsSet} customizer
+	 */
+	public void claims(Consumer<Map<String, Object>> claimsCustomizer) {
```

Comment:
Let's change to `Consumer<JwtClaimsSet.Builder>`

Reply:
That's how I did it initially. But to be fair, now I like map as an argument more for a couple of reasons. Firstly, the given consumer goes directly into `JwtClaimsSet.Builder.claims` method, and there is no need to do something like `tokenBuilder.claims(builder -> builder.claims(claims -> claims.get("claim-name")))` to access existing claims. Secondly, map is a abstract enough and even if underlying `JwtClaimsSet` implementation changes, there'll be no need for users to change their code. `JwtBuilder` can even decide to switch implementation from `JwtClaimsSet` completely without affecting user's code.

If it's not convincing, I'm happy to change it to `JwtClaimsSet.Builder` but in this case there is no need in the consumer, `JwtTokenBuilder` can simply return claims set builder as `JwtClaimsSet.Builder getClaimsBuilder() { return this.jwtClaimsSetBuilder; }`.